build.assets/robotest/test.sh

```diff
@@ -0,0 +1,50 @@
+#!/bin/bash
+
+# This file contains unit tests for the upgrade recommendation algorithm.
+#
+# None of the values in this file need to be kept up to date as new releases
+# come out, because the algorithm should never recommend a release that is
+# chronologically more recent than the release being upgraded to.  Because all
+# test cases use a specific tag (e.g. 7.0.10) for "to" instead of a floating
+# ref (e.g. 7.0.x) the recommended "from" releases should be fixed, regardless of
+# new releases on the same branch.
+set -o errexit
+set -o nounset
+set -o pipefail
+set -o xtrace
+
+source $(dirname $0)/utils.sh
+
+# Check recommended upgrade where the from branch has more current etcd
+# versions than the to release.
+#
+# 7.0.0 supports upgrade from etcd 3.3.12 as its most current 3.3.
+# 6.1.25 bumps etcd to 3.3.20, meaning 6.1.24 is the last 6.1.x that can upgrade
+# to 7.0.0.
+#
+# See https://github.com/gravitational/gravity/issues/1735 for more info.
+#
+# Due to difficulty of determining etcd compatibility, this test uses 6.1.20 as a
+# conservative stand-in for 6.1.24 because 6.1.20 the last 6.1.x that came out before 7.0.0.
+if [[ "$(recommended_upgrade_tag_between $(branch 6.1.x) 7.0.0)" != "6.1.20" ]];  then
+  exit 1
+fi
+
+# Check a more current 7.0.x upgrades from a more current 6.1.x.
+#
+# 7.0.11 supports upgrade from etcd 3.3.22, which is what 6.1.29 uses.
+if [[ "$(recommended_upgrade_tag_between $(branch 6.1.x) 7.0.11)" != "6.1.29" ]];  then
+  exit 1
+fi
+
+# Check that an upgrade from a release to itself is not recommended.
+if [[ "$(recommended_upgrade_tag_between $(branch 7.0.x) 7.0.10)" != "7.0.9" ]];  then
+  exit 1
+fi
+
+# Check recommended upgrade when patch-1  doesn't exist.
+#
+# 6.1.23 is the only recent case of this in the wild.
```

Comment:
So, I'm still unsure how this works. If you wanted to skip `6.1.23` -> `6.1.24` upgrade path, why is it not excluding `6.1.23` explicitly and instead uses a condition to avoiding anything other than `6.1.22`?
I guess my question is - how is this going to be used and when would this error (eg `exit 1`) be triggered?

Reply:
> why is it not excluding 6.1.23 explicitly and instead uses a condition to avoiding anything other than 6.1.22?

Perhaps it would be easier to read if the condition wasn't negated?

```bash
if [[ "$(recommended_upgrade_tag_between $(branch 6.1.x) "6.1.24")" == "6.1.22" ]];  then
  echo "Test passed"
else
  exit 1
fi
```

Or are you suggesting:

```bash
if [[ "$(recommended_upgrade_tag_between $(branch 6.1.x) "6.1.24")" == "6.1.23" ]];  then
  exit 1
else
  echo "Test passed"
fi
```

The second test case allows for other silliness like recommending 6.1.24.  So while it is more pure, it is less useful as a test IMO.

> I guess my question is - how is this going to be used and when would this error (eg exit 1) be triggered?

These unit tests are used used by calling the file directly:

```
walt@work:~/git/gravity$ ./build.assets/robotest/test.sh 
++ dirname ./build.assets/robotest/test.sh    
+ source ./build.assets/robotest/utils.sh                                                      
++ set -o errexit                                                                                                                                                                             
++ set -o nounset                                                                              
++ set -o pipefail                 
+++ branch 6.1.x                                                                               
+++ local branch_prefix=remotes/grav/version
...
```

Exit 1 is triggered if any test cases fails.  I'd expect that following exit 1, the employee who triggered the test would look at why test is there and if recommended_upgrade_tag_between has changed.

This file is not presently integrated with `make test` or any make targets, so it won't be run by any CI.  I could probably clean it up (turn off `xtrace`, add some logging statements) and integrate it.  However my my goal I wasn't to invent or integrate a bash test harness.  I am trying to make sure useful builds were recommended in these situations, and writing the these tests seemed like the easiest way to do so.

As a second goal. these tests serve as concrete documenation for what the algorithm needs to handle.  If/when someone re-implements a better version of this (outside of bash? maybe ditching git tags as a strategy?), the scenarios these test cases check should still be relevant.

I'd considered moving to a better language myself, but go -> bash (robotest's `run_suite.sh`) -> go felt more byzantine than the current bash configuration, and I didn't have time/priority to translate all of robotest's bindings into native go.

Reply:
Because:

1) The only remaining discussion point is test code that does not affect execution
2) I've been iterating on this PR for 9 days

I'm going to go ahead and merge as is.

That being said, I do want to get to the bottom of this -- I'll just handle any additional changes in a follow up PR.

I see ~4 reasonable paths:

1) leave it as is (if I've convinced you in the discussion here)
2) refactor test cases to be more clear (see https://github.com/gravitational/gravity/pull/1837#discussion_r452948904)
3) option 2 + integrate tests with make/CI such that they're regularly run
4) remove the unit tests entirely.... They cause more confusion than they prevent.

I prefer one of 1-3, and I think 2 is closest to what you're looking for.